uarray/_uarray_dispatch.cxx

```diff
@@ -0,0 +1,491 @@
+
+#include <Python.h>
+#include <utility>
+#include <new>
+#include <cstddef>
+
+namespace {
+class py_ref
+{
+	explicit py_ref(PyObject * object): obj(object) {}
+public:
+
+	py_ref(): obj(nullptr) {}
+	py_ref(std::nullptr_t): py_ref() {}
+
+	py_ref(const py_ref & other): obj(other.obj) { Py_XINCREF(obj); }
+	py_ref(py_ref && other): obj(other.obj) { other.obj = nullptr; }
+
+	static py_ref steal(PyObject * object) { return py_ref(object); }
+
+	static py_ref ref(PyObject * object)
+		{
+			Py_XINCREF(object);
+			return py_ref(object);
+		}
+
+	~py_ref(){ Py_XDECREF(obj); }
+
+	py_ref & operator = (const py_ref & other)
+		{
+			py_ref(other).swap(*this);
```

Comment:
Is this right? Shouldn't you just be copying and `INCREF`-ing `other.obj`?

Reply:
It's the [Copy-Swap idiom](https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Copy-and-swap).
The temporary is copy constructed (which calls `Py_XINCREF`) and then the pointer is swapped into place. When the temporary is destroyed, the old value is `Py_XDECREF`ed.